README.txt
DNSMAP README FILE

INTRODUCTION

dnsmap was originally released back in 2006 and was inspired by the
fictional story "The Thief No One Saw" by Paul Craig, which can be found
in the book "Stealing the Network - How to 0wn the Box"

dnsmap is mainly meant to be used by pentesters during the information
gathering/enumeration phase of infrastructure security assessments. During the
enumeration stage, the security consultant would typically discover the target
company's IP netblocks, domain names, phone numbers, etc ...

Subdomain brute-forcing is another technique that should be used in the
enumeration stage, as it's especially useful when other domain enumeration
techniques such as zone transfers don't work (I rarely see zone transfers
being publicly allowed these days by the way).

If you are interested in researching stealth computer intrusion techniques,
I suggest reading this excellent (and fun) chapter which you can find for
*free* on the web:

http://www.ethicalhacker.net/content/view/45/2/

I'm happy to say that dnsmap was included in Backtrack 2 and 3 - although the version
included is now quite dated - and has been reviewed by the community:

http://backtrack.offensive-security.com/index.php?title=Tools
http://www.linuxhaxor.net/2007/07/14/backtrack-2-information-gathering-all-dnsmap/
http://www.darknet.org.uk/2009/03/dnsmap-022-released-subdomain-bruteforcing-tool/
http://www.gnucitizen.org/blog/new-version-of-dnsmap-out/


COMPILING

Compiling should be straightforward:

$ make

Or:

$ gcc -Wall dnsmap.c -o dnsmap


INSTALLATION

Example of manual installation:

# cp ./dnsmap /usr/local/bin/dnsmap

If you wish to bruteforce several target domains in bulk fashion, you can use the
included dnsmap-bulk.sh script. Just copy the script to /usr/local/bin/ so you can 
call it from any location. i.e.:

# cp ./dnsmap-bulk.sh /usr/local/bin/

And set execute permissions. e.g.:

# chmod ugo+x /usr/local/bin/dnsmap-bulk.sh


LIMITATIONS

This tool won't work with target domains which use wildcards. When a domain
uses wildcards, all bruteforced subdomains will resolve to the same IP address,
which makes enumerating target servers unfeasible.

dnsmap *does* however inform the user when wildcards have been detected and suggests
choosing a different target domain.


FUN THINGS THAT CAN HAPPEN

1. Finding interesting remote access servers (i.e.: https://extranet.targetdomain.com)

2. Finding badly configured and/or unpatched servers (i.e.: test.targetdomain.com)

3. Finding new domain names which will allow you to map non-obvious/hard-to-find netblocks
   of your target organization (registry lookups - aka whois is your friend)

4. Sometimes you find that some bruteforced subdomains resolve to internal IP addresses
   (RFC 1918). This is great as sometimes they are real up-to-date "A" records which means that
   it *is* possible to enumerate internal servers of a target organization from the Internet
   by only using standard DNS resolving (as oppossed to zone transfers for instance).


USAGE

Bruteforcing can be done either with dnsmap's built-in wordlist or a user-supplied wordlist.
Results can be saved in CSV and human-readable format for further processing. dnsmap does NOT
require root privileges to be run, and should NOT be run with such privileges for security reasons.

The usage syntax can be obtained by simply running dnsmap without any parameters:

$ ./dnsmap

dnsmap 0.24 - DNS Network Mapper by pagvac (gnucitizen.org)

usage: dnsmap <target-domain> [options]
options:
-w <wordlist-file>
-r <regular-results-file>
-c <csv-results-file>
-d <delay-milliseconds>
e.g.:
dnsmap target-domain.foo
dnsmap target-domain.foo -w yourwordlist.txt -r /tmp/domainbf_results.txt
dnsmap target-fomain.foo -r /tmp/ -d 3000
dnsmap target-fomain.foo -r ./domainbf_results.txt

Example of subdomain bruteforcing using dnsmap's built-in word-list:

$ ./dnsmap targetdomain.foo

Example of subdomain bruteforcing using a user-supplied wordlist:

$ ./dnsmap targetdomain.foo -w wordlist.txt

Example of subdomain bruteforcing using the built-in wordlist and saving the results to /tmp/ :

$ ./dnsmap targetdomain.foo -r /tmp/

Since no filename was provided in the previous example, but rather only a path, dnsmap would create an unique filename which includes the current timestamp. E.g.: /tmp/dnsmap_targetdomain_foo_2009_12_15_234953.txt

Example of subdomain bruteforcing using the built-in wordlist, saving the results to /tmp/, and waiting a random maximum of 3 milliseconds between each request:

$ ./dnsmap targetdomain.foo -r /tmp/ -d 3

It is recommended to use the -d (delay in milliseconds) option in cases where dnsmap is interfering with your online experience. i.e.: killing your bandwidth

For bruteforcing a list of target domains in a bulk fashion use the bash script provided. e.g.:

$ ./dnsmap-bulk.sh domains.txt /tmp/results/


WORDLISTS

http://packetstormsecurity.org/Crackers/wordlists/dictionaries/
http://www.cotse.com/tools/wordlists1.htm


OTHER SIMILAR TOOLS - choice is freedom!

WS-DNS-BFX
http://ws.hackaholic.org/tools/WS-DNS-BFX.tgz

DNSDigger
http://www.ernw.de/download/dnsdigger.zip

Fierce Domain Scan
http://ha.ckers.org/fierce/

Desperate
http://www.sensepost.com/research_misc.html

DNSenum
http://dnsenum.googlecode.com/files/dnsenum1.2.tar.gz

ReverseRaider
http://complemento.sourceforge.net/


--
pagvac | GNUCITIZEN.org
January 2010


Changelog.txt
XX/01/2010: dnsmap 0.24 released at www.gnucitizen.org
* IPv6 support
* Makefile included
* delay option (-d) added. This is useful in cases where dnsmap is killing your bandwidth
* changes made to make dnsmap compatible with OpenDNS
* disclosure of internal IP addresses (RFC 1918) are reported
* updated built-in wordlist
* domains susceptible to "same site" scripting (http://snipurl.com/etbcv) are reported
* completion time is now displayed to the user
* mechanism to attempt to bruteforce wildcard-enabled domains
* unique filename containing timestamp is now created when no specific output filename is supplied by user
* various minor bugs fixed

22/02/2009: dnsmap 0.22 released at www.gnucitizen.org
* saving the results in human-readable and CSV format for easy processing
* fixed bug that disallowed reading wordlists with DOS CRLF format
* improved built-in subdomains wordlist
* new bash script (dnsmap-bulk.sh) included which allows running dnsmap against a list of domains from a user-supplied file. i.e.: bruteforcing several domains in a bulk fashion
* bypassing of signature-based dnsmap detection by generating a proper pseudo-random subdomain when checking for wildcards

17/08/2006: dnsmap 0.1 (first public version) released at foro.elhacker.net
* bruteforcing based on builtin list and user-supplied wordlist
* obtain all available A records for each bruteforced (sub)domain (rather than only one)


dnsmap.h
/*
 * ** dnsmap - DNS Network Mapper by pagvac
 * ** Copyright (C) 2009 gnucitizen.org
 * **
 * ** This program is free software; you can redistribute it and/or modify
 * ** it under the terms of the GNU General Public License as published by
 * ** the Free Software Foundation; either version 2 of the License, or
 * ** (at your option) any later version.
 * **
 * ** This program is distributed in the hope that it will be useful,
 * ** but WITHOUT ANY WARRANTY; without even the implied warranty of
 * ** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * ** GNU General Public License for more details.
 * **
 * ** You should have received a copy of the GNU General Public License
 * ** along with this program; if not, write to the Free Software
 * ** Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 * */

#define MAXSUBSIZE 	100
#define MAXSTRSIZE 	200
//#define MAXIPSIZE	16
#define BANNER 		"dnsmap 0.24 - DNS Network Mapper by pagvac (gnucitizen.org)\n\n"
#define USAGE 		"usage: dnsmap <target-domain> [options]\noptions:\n"\
			"-w <wordlist-file>\n-r <regular-results-file>\n-c <csv-results-file>\n"\
			"-d <delay-milliseconds>\n"
#define EXAMPLES 	"e.g.:\ndnsmap target-domain.foo\n"\
			"dnsmap target-domain.foo -w yourwordlist.txt -r /tmp/domainbf_results.txt\n"\
			"dnsmap target-fomain.foo -r /tmp/ -d 3000\n"\
			"dnsmap target-fomain.foo -r ./domainbf_results.txt\n\n"
#define INTIPWARN	"[+] warning: internal IP address disclosed\n"
#define SAMESITEXSSWARN "[+] warning: target domain might be vulnerable to \"same site\" scripting (http://snipurl.com/etbcv)\n"
#define WILDCARDWARN	"[+] warning: the target domain might use wildcards. "\
			"dnsmap will try to filter out false positives\n"
#define INPUTERR	"[+] error: entered parameter(s) is/are too long!\n"
#define DELAYINPUTERR	"[+] error: delay must be at least 1 second!\n"
#define DOMAINERR	"[+] error: entered domain is not valid!\n"
#define CREATEFILEERR	"%s\"%s\"!\n\n", "[+] error creating results file on ", argv[(i+1)]
#define OPENFILEERR	"%s\"%s\"!\n\n", "[+] error opening wordlist file ", wordlistFilename
#define OPENDNSMSG	"[+] openDNS detected. good! this might help with performance\n"
#define BUILTINMSG	"%s%s%s\n", "[+] searching (sub)domains for ", argv[1], " using built-in wordlist"
#define EXTERNALMSG	"%s%s%s%s\n", "[+] searching (sub)domains for ", argv[1], " using ", wordlistFilename
#define DELAYMSG	"%s%d%s\n", "[+] using maximum random delay of ", milliseconds, " millisecond(s) between requests"
#define RESULTSMSG1	"[+] %d internal IP addresses disclosed\n", intIPcount
#define RESULTSMSG2	"[+] regular-format results can be found on %s\n", txtResultsFilename
#define RESULTSMSG3	"[+] completion time: %lu second(s)\n", end-start
#define RESULTSMSG4	"[+] %d%s%d%s\n", found, " (sub)domains and ",ipCount, " IP address(es) found"
#define RESULTSMSG5	"[+] csv-format results can be found on %s\n", csvResultsFilename
#define FALSE 0
#define TRUE 1
#define DEBUG 0

// buil-in list of subdomains
// mainly targeting English and Spanish keywords
char sub[][MAXSUBSIZE]=
{
"access",
"accounting",
"accounts",
"admin",
"administrator",
"aix",
"ap",
"archivos",
"aula",
"aulas",
"ayuda",
"backup",
"backups",
"bart",
"bd",
"beta",
"biblioteca",
"billing",
"blackboard",
"blog",
"blogs",
"bsd",
"carro",
"cart",
"cas",
"catalog",
"catalogo",
"catalogue",
"chat",
"chimera",
"citrix",
"classroom",
"clientes",
"clients",
"connect",
"controller",
"correoweb",
"cpanel",
"csg",
"customers",
"data",
"db",
"dbs",
"dc",
"demo",
"demon",
"demostration",
"descargas",
"developers",
"development",
"diana",
"directory",
"dmz",
"domain",
"domain-controller",
"domaincontroller",
"download",
"downloads",
"ds",
"eaccess",
"ejemplo",
"ejemplos",
"email",
"enrutador",
"eventos",
"events",
"example",
"examples",
"exchange",
"extranet",
"files",
"finance",
"firewall",
"foro",
"foros",
"forum",
"forums",
"freebsd",
"ftp",
"ftpd",
"fw",
"galeria",
"gallery",
"gateway",
"gilford",
"groups",
"groupwise",
"guest",
"guia",
"guide",
"gw",
"help",
"helpdesk",
"hera",
"heracles",
"hercules",
"home",
"homer",
"hotspot",
"hr",
"hypernova",
"images",
"imail",
"imap",
"imap3",
"imap3d",
"imapd",
"imaps",
"imgs",
"imogen",
"inmuebles",
"internal",
"interno",
"intranet",
"ipsec",
"ipv6",
"irc",
"ircd",
"jabber",
"lab",
"laboratories",
"laboratorio",
"labs",
"library",
"linux",
"lisa",
"localhost",
"log",
"login",
"logon",
"logs",
"mail",
"mailgate",
"manager",
"marketing",
"member",
"members",
"mercury",
"meta",
"meta01",
"meta02",
"meta03",
"meta1",
"meta2",
"meta3",
"miembros",
"minerva",
"mob",
"mobile",
"moodle",
"movil",
"mssql",
"mx",
"mx0",
"mx01",
"mx02",
"mx03",
"mx1",
"mx2",
"mx3",
"mysql",
"nelson",
"neon",
"net",
"netmail",
"news",
"novell",
"ns",
"ns0",
"ns01",
"ns02",
"ns03",
"ns1",
"ns2",
"ns3",
"ntp",
"online",
"ora",
"oracle",
"osx",
"owa",
"partners",
"pcanywhere",
"pegasus",
"pendrell",
"personal",
"photo",
"photos",
"pop",
"pop3",
"portal",
"postgresql",
"postman",
"postmaster",
"pp", // preprod?
"ppp",
"pre-prod",
"preprod",
"private",
"prod",
"proxy",
"prueba",
"pruebas",
"pub",
"public",
"ras",
"remote",
"reports",
"research",
"restricted",
"robinhood",
"router",
"rtr",
"sales",
"sample",
"samples",
"sandbox",
"search",
"secure",
"seguro",
"server",
"services",
"servicios",
"servidor",
"sharepoint",
"shop",
"shopping",
"sms",
"smtp",
"socios",
"solaris",
"soporte",
"sp", // sharepoint?
"sql",
"squirrel",
"squirrelmail",
"ssh",
"staff",
"staging",
"stats",
"sun",
"support",
"test",
"tftp",
"tienda",
"uat",
"unix",
"upload",
"uploads",
"ventas",
"virtual",
"vista",
"vm",
"vnc",
"vpn",
"vpn1",
"vpn2",
"vpn3",
"wap",
"web",
"web1",
"web2",
"web3",
"webadmin",
"webct",
"weblog",
"webmail",
"webmaster",
"webmin",
"win",
"windows",
"ww0",
"ww01",
"ww02",
"ww03",
"ww1",
"ww2",
"ww3",
"www",
"www0",
"www01",
"www02",
"www03",
"www1",
"www2",
"www3",
"xanthus"
};
